README
NAME
    Algorithm::MTF - Move-to-front transform encoder/decoder

SYNOPSIS
      use Algorithm::MTF;

      my $encoder = Algorithm::MTF::Encoder->new;
      my $code = $mtf->encode("aaabac"); ## $code is [ 97, 0, 0, 98, 1, 99 ]

      my $decoder = Algorithm::MTF::Decoder->new;
      say $mtf->decode([ 97, 0, 0, 98, 1, 99 ]); # "aaabac"

DESCRIPTION
    The move-to-front (or MTF) transform is an encoding of data (typically a
    stream of bytes) designed to improve the performance of entropy encoding
    techniques of compression. When efficiently implemented, it is fast
    enough that its benefits usually justify including it as an extra step
    in data compression algorithms. (quote from en.wikipedia.org)

SEE ALSO
    <http://en.wikipedia.org/wiki/Move-to-front_transform>

AUTHOR
    Naoya Ito, <naoya at bloghackers.net<gt>

COPYRIGHT AND LICENSE
    Copyright (C) 2008 by Naoya Ito

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8.8 or, at
    your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension Algorithm::MTF.

0.01  Sun Oct 19 17:14:14 2008
	- original version; created by h2xs 1.23 with options
		-AX -n Algorithm::MTF



bench.pl
#!/usr/bin/env perl
use strict;
use warnings;
use FindBin::libs;

use Perl6::Say;
use Algorithm::DivSufSort;
use Benchmark;
use Path::Class qw/file/;

use Algorithm::MTF;
use Algorithm::MTF::List;
use Algorithm::MTF::Array;

sub bs_encode ($) {
    my $text = shift;
    my $sa = divsufsort( $text );

    my @text = unpack('C*', $text);
    return join '', map { chr $text[$_ -1] } @$sa;
}

my $bwt = bs_encode file('/etc/httpd/conf/mime.types')->slurp;

my $list = Algorithm::MTF::List->new;
for (my $i = 0xff; $i >= 0; $i--) {
    $list->insert(chr $i);
    # $list->insert( Algorithm::MTF::List::Node->new(chr $i) );
}
my $list_mtf = Algorithm::MTF::Encoder->new($list);

my $array = Algorithm::MTF::Array->new;
for (my $i = 0; $i < 0x100; $i++) {
    $array->[$i] = chr $i;
}
my $array_mtf = Algorithm::MTF::Encoder->new($array);

timethese(10, {
    list  => sub {
        $list_mtf->encode($bwt);
    },
    # array => sub {
    #    $array_mtf->encode($bwt);
    # },
});


lib/Algorithm/MTF/List.pm
package Algorithm::MTF::List;
use strict;
use warnings;
use base qw/Class::Accessor::Lvalue::Fast/;

use Algorithm::MTF::List::Node;

__PACKAGE__->mk_accessors(qw/first/);

sub new {
    my $class = shift;
    my $self = $class->SUPER::new(@_);

    ## the first element is a sentinel.
    $self->first = Algorithm::MTF::List::Node->new;

    return bless $self, $class;
}

sub insert {
    my ($self, $node) = @_;

    if (not ref $node) {
        $node = Algorithm::MTF::List::Node->new($node);
    }

    $self->first->insert_next($node);
}

sub mtf {
    my ($self, $value) = @_;

    my $node = $self->first;
    my $pos = 0;

    while (defined $node->next) {
        if ($node->next->value eq $value) {
            last;
        }
        $pos++;
        $node = $node->next;
    }

    if ($pos > 0) {
        $self->insert( $node->remove_next );
    }

    return $pos;
}

sub mtf_by_pos {
    my ($self, $pos) = @_;

    my $node = $self->first;
    for (my $i = 0; $i < $pos; $i++) {
        $node = $node->next;
    }

    my $value = $node->next->value;

    if ($pos > 0) {
        $self->insert( $node->remove_next );
    }

    return $value;
}

1;


lib/Algorithm/MTF/Array.pm
package Algorithm::MTF::Array;
use strict;
use warnings;

sub new {
    my $class = shift;
    bless [] ,$class;
}

sub mtf {
    my ($self, $value) = @_;
    my $i = $self->index_of( $value );
    if ($i > 0) {
        $self->delete_at($i);
        unshift @$self, $value;
    }
    return $i;
}

sub mtf_by_pos {
    my ($self, $pos) = @_;
    my $c = $self->[$pos];
    if ($pos > 0) {
        $self->delete_at($pos);
        unshift @$self, $c;
    }
    return $c;
}

sub delete_at {
    my ($self, $i) = @_;
    if ($i > @$self - 1) {
        return;
    }

    my @end = splice @$self, $i;
    my $value = shift @end;
    push @$self, @end;

    return $value;
}

sub index_of {
    my ($self, $c) = @_;
    for (my $i = 0; $i < @$self; $i++) {
        if ($self->[$i] eq $c) {
            return $i;
        }
    }
}

1;


lib/Algorithm/MTF.pm
package Algorithm::MTF;
use strict;
use warnings;
use base qw/Class::Accessor::Lvalue::Fast/;

our $VERSION = '0.01';

__PACKAGE__->mk_accessors(qw/table/);

use UNIVERSAL::require;
use Algorithm::MTF::Encoder;
use Algorithm::MTF::Decoder;

sub new {
    my ($class, $backend) = @_;
    my $self = $class->SUPER::new;

    if (not defined $backend) {
        Algorithm::MTF::List->require or die $@;
        my $list = Algorithm::MTF::List->new;
        for (my $i = 0xff; $i >= 0; $i--) {
            $list->insert(chr $i);
        }
        $self->table = $list;

#         Algorithm::MTF::Array->require or die $@;
#         my $array = Algorithm::MTF::Array->new;
#         for (my $i = 0; $i < 0x100; $i++) {
#             $array->[$i] = chr $i;
#         }
#        $self->table = $array;
    } else {
        $self->table = $backend;
    }

    bless $self, $class;
}

1;

__END__

=head1 NAME

Algorithm::MTF - Move-to-front transform encoder/decoder

=head1 SYNOPSIS

  use Algorithm::MTF;

  my $encoder = Algorithm::MTF::Encoder->new;
  my $code = $mtf->encode("aaabac"); ## $code is [ 97, 0, 0, 98, 1, 99 ]

  my $decoder = Algorithm::MTF::Decoder->new;
  say $mtf->decode([ 97, 0, 0, 98, 1, 99 ]); # "aaabac"

=head1 DESCRIPTION

The move-to-front (or MTF) transform is an encoding of data (typically
a stream of bytes) designed to improve the performance of entropy
encoding techniques of compression. When efficiently implemented, it
is fast enough that its benefits usually justify including it as an
extra step in data compression algorithms. (quote from en.wikipedia.org)

=head1 SEE ALSO

L<http://en.wikipedia.org/wiki/Move-to-front_transform>

=head1 AUTHOR

Naoya Ito, E<lt>naoya at bloghackers.net<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 by Naoya Ito

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut
